autoarray/structures/grids.py

```diff
@@ -1555,7 +1715,172 @@ def extent(self):
+
+        Parameters
+        ----------
+        unmasked_sparse_shape : (int, int)
+            The shape of the unmasked sparse-grid whose centres form the sparse-grid.
+        pixel_scales : (float, float)
+            The pixel conversion scale of a pixel in the y and x directions.
+        grid : Grid
+            The grid used to determine which pixels are in the sparse grid.
+        origin : (float, float)
+            The centre of the unmasked sparse grid, which matches the centre of the mask.
+        """
+        self.sparse = sparse_grid
+        self.sparse_1d_index_for_mask_1d_index = sparse_1d_index_for_mask_1d_index
+
+    @classmethod
+    def from_grid_and_unmasked_2d_grid_shape(cls, grid, unmasked_sparse_shape):
+        """Calculate the image-plane pixelization from a grid of coordinates (and its mask).
+
+        See *grid_stacks.GridSparse* for details on how this grid is calculated.
+
+        Parameters
+        -----------
+        grid : grids.Grid
+            The grid of (y,x) scaled coordinates at the centre of every image value (e.g. image-pixels).
+        """
+
+        pixel_scales = grid.mask.pixel_scales
+
+        pixel_scales = (
+            (grid.shape_2d_scaled[0] + pixel_scales[0]) / (unmasked_sparse_shape[0]),
+            (grid.shape_2d_scaled[1] + pixel_scales[1]) / (unmasked_sparse_shape[1]),
+        )
+
+        origin = grid.geometry.mask_centre
+
+        unmasked_sparse_grid_1d = grid_util.grid_1d_via_shape_2d(
+            shape_2d=unmasked_sparse_shape,
+            pixel_scales=pixel_scales,
+            sub_size=1,
+            origin=origin,
+        )
+
+        unmasked_sparse_grid_pixel_centres = grid_util.grid_pixel_centres_1d_from_grid_scaled_1d_shape_2d_and_pixel_scales(
+            grid_scaled_1d=unmasked_sparse_grid_1d,
+            shape_2d=grid.mask.shape,
+            pixel_scales=grid.mask.pixel_scales,
+        ).astype(
+            "int"
+        )
+
+        total_sparse_pixels = mask_util.total_sparse_pixels_from_mask_2d(
+            mask_2d=grid.mask,
+            unmasked_sparse_grid_pixel_centres=unmasked_sparse_grid_pixel_centres,
+        )
+
+        sparse_for_unmasked_sparse = sparse_util.sparse_for_unmasked_sparse_from_mask_2d_and_pixel_centres(
+            mask_2d=grid.mask,
+            unmasked_sparse_grid_pixel_centres=unmasked_sparse_grid_pixel_centres,
+            total_sparse_pixels=total_sparse_pixels,
+        ).astype(
+            "int"
+        )
+
+        unmasked_sparse_for_sparse = sparse_util.unmasked_sparse_for_sparse_from_mask_2d_and_pixel_centres(
+            total_sparse_pixels=total_sparse_pixels,
+            mask_2d=grid.mask,
+            unmasked_sparse_grid_pixel_centres=unmasked_sparse_grid_pixel_centres,
+        ).astype(
+            "int"
+        )
+
+        regular_to_unmasked_sparse = grid_util.grid_pixel_indexes_1d_from_grid_scaled_1d_shape_2d_and_pixel_scales(
```

Comment:
These function names are a too long.

You can actually require usage of keyword arguments thereby avoiding the need for the "from" part of the function

```python
def grid_pixel_indexes_1d(
    *,
    grid_scaled_1d=...
)
```

Of course this only works if you don't get a clash in function names

Reply:
Adpoted this convention, but use of * is restricted due to numba and values with non default values.